lib/maths/analytics/unittest/CDataFrameUtilsTest.cc

```diff
@@ -802,6 +802,133 @@ BOOST_AUTO_TEST_CASE(testStratifiedSamplingRowMasks) {
     }
 }
 
+BOOST_AUTO_TEST_CASE(testDistributionPreservingSamplingRowMasks) {
+    using TDoubleDoubleUMap = boost::unordered_map<double, double>;
+    test::CRandomNumbers testRng;
+    maths::common::CPRNG::CXorOShiro128Plus rng;
+
+    TSizeVec numberDistributionSourceRows(1);
+    std::size_t numberAdditionalRows{1000};
+    std::size_t numberCols{1};
+
+    // Test that for categorical data the count of individual classes remains the same.
+    for (std::size_t trial = 0; trial < 10; ++trial) {
+        testRng.generateUniformSamples(500, 750, 1, numberDistributionSourceRows);
+        TDoubleVec categories;
+        testRng.generateNormalSamples(0.0, 3.0, numberDistributionSourceRows[0], categories);
+
+        auto frame = core::makeMainStorageDataFrame(numberCols).first;
+        frame->categoricalColumns(TBoolVec{true});
+        for (std::size_t i = 0; i < numberDistributionSourceRows[0]; ++i) {
+            frame->writeRow([&](core::CDataFrame::TFloatVecItr column, std::int32_t&) {
+                *column = std::floor(std::fabs(categories[i]));
+            });
+        }
+        for (std::size_t i = 0; i < numberAdditionalRows; ++i) {
+            frame->writeRow([&](core::CDataFrame::TFloatVecItr column,
+                                std::int32_t&) { *column = 0; });
+        }
+        frame->finishWritingRows();
+        core::CPackedBitVector distributionSourceRowsMask{
+            numberDistributionSourceRows[0], true};
+        distributionSourceRowsMask.extend(false, numberAdditionalRows);
+
+        core::CPackedBitVector allRowsMask{distributionSourceRowsMask.size(), true};
+
+        TDoubleDoubleUMap expectedCategoryCounts;
+        for (auto i = distributionSourceRowsMask.beginOneBits();
+             i != distributionSourceRowsMask.endOneBits(); ++i) {
+            expectedCategoryCounts[std::floor(std::fabs(categories[*i]))] += 1.0;
+        }
+
+        auto samplingRowsMask = maths::analytics::CDataFrameUtils::distributionPreservingSamplingRowMask(
+            1, *frame, 0, rng, numberDistributionSourceRows[0], 10,
+            distributionSourceRowsMask, allRowsMask);
+
+        BOOST_REQUIRE_EQUAL(samplingRowsMask.size(), allRowsMask.size());
+        BOOST_REQUIRE_CLOSE(samplingRowsMask.manhattan(),
+                            numberDistributionSourceRows[0], 1);
+
+        TDoubleDoubleUMap actualCategoryCounts;
+        frame->readRows(1, 0, frame->numberRows(),
+                        [&](const core::CDataFrame::TRowItr& beginRows,
+                            const core::CDataFrame::TRowItr& endRows) {
+                            for (auto row = beginRows; row != endRows; ++row) {
+                                actualCategoryCounts[(*row)[0]] += 1.0;
+                            }
+                        },
+                        &samplingRowsMask);
```

Comment:
Maybe factor this out into a lambda and use for computing actual and expected counts. Alternatively, I guess this could use CDataFrameUtils::categoryCounts, but maybe testing via an alternative implementation is better.

Reply:
I refactored it to use `CDataFrameUtils::categoryCounts` for `actualCategoryCounts` and I use explicit calculation for `expectedCategoryCounts`. This way I can test the `categoryCounts` implementation and keep the test code clean.